### PR TITLE
Improve handling of public keys

### PR DESCRIPTION
**Pull request by simo5**
#### Description

This PR improves pkcs11-provider's handling of public keys in three related and interconnected areas by creating a first class association between private and public keys.

In OpenSSL, the EVP_PKEY type can carry both public and private key information in a single structure, and this is reflected sometimes heavily in the APIs and application usge of this type. The most glaring case is key pair generation.

To better handle cases where a private key i sourced but information about the public key is requested we allow a private key object to carry a refcounted pointer to the corresponding public key object.

The three main changes in this PR are:
- key generation, we stop copying public attributes on the private key object on generation, and instead we associated the public key to the private key we return to the caller.
- EC public key import is fixed to properly handle this behavior change, in the process the whole EC import/export code has been improved to fix broken corner cases
- Public Key output, encoder functions to print public key information have been fixed to find the associate public key when public key info is requested for a private key.

A new pkey test has been added to check basic keygen and printing aspects, the existing CMS and ECDH tests fully exercise the EC public key export/import changes, and the edwards tests exercise the ED export changes.

#### Checklist

- [x] Code modified for feature
- [x] Test suite updated with functionality tests
- [x] Test suite updated with negative tests
- ~[ ] Documentation updated~

#### Reviewer's checklist:

- [ ] Any issues marked for closing are addressed
- [ ] There is a test suite reasonably covering new functionality or modifications
- [ ] This feature/change has adequate documentation added
- [ ] Code conform to coding style that today cannot yet be enforced via the check style test
- [ ] Commits have short titles and sensible commit messages
- [ ] Coverity Scan has run if needed (code PR) and no new defects were found

### CODE REVIEW

**Review by Jakuje — APPROVED** [review]

File: tests/tpkey.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
+/* Copyright (C) 2022 Simo Sorce <simo@redhat.com>
```
[the remark this answers is not in the document]
Comment (simo5) [reply tests/tpkey.c:1]: heh

**Comment by simo5** [comment]
Thanks for the review @Jakuje
